src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -931,29 +931,39 @@ protected Object matchInRegionTRegex(
             int fromIndex = fromPos;
             final Object interopByteArray;
             final String execMethod;
-            if (startPosNotZeroProfile.profile(startPos > 0)) {
-                // GR-32765: When adopting TruffleString, use a TruffleString substring here instead
-                // If startPos != 0, then fromPos == startPos.
-                assert fromPos == startPos;
-                fromIndex = 0;
 
-                if (getBytesObjectNode == null) {
-                    CompilerDirectives.transferToInterpreterAndInvalidate();
-                    getBytesObjectNode = insert(RopeNodes.GetBytesObjectNode.create());
+            if (createMatchDataProfile.profile(createMatchData)) {
+                if (startPosNotZeroProfile.profile(startPos > 0)) {
+                    // GR-32765: When adopting TruffleString, use a TruffleString substring here instead
+                    // If startPos != 0, then fromPos == startPos.
+                    assert fromPos == startPos;
+                    fromIndex = 0;
+
+                    if (getBytesObjectNode == null) {
+                        CompilerDirectives.transferToInterpreterAndInvalidate();
+                        getBytesObjectNode = insert(RopeNodes.GetBytesObjectNode.create());
+                    }
+                    interopByteArray = getBytesObjectNode.getRange(rope, startPos, toPos);
+                    execMethod = "exec";
+                } else {
+                    final byte[] bytes = bytesNode.execute(rope);
+                    interopByteArray = getContext().getEnv().asGuestValue(bytes);
+                    execMethod = "execBytes";
                 }
-                interopByteArray = getBytesObjectNode.getRange(rope, startPos, toPos);
-                execMethod = "exec";
             } else {
+                // Only strscan ever passes a non-zero startPos and that never uses `match?`.
+                assert startPos == 0 : "Simple Boolean match not supported with non-zero startPos";
+
                 final byte[] bytes = bytesNode.execute(rope);
                 interopByteArray = getContext().getEnv().asGuestValue(bytes);
-                execMethod = "execBytes";
+                execMethod = "execBoolean";
```

Comment:
It doesn't accept a `byte[]` array, only `TruffleString` or `String`.

Reply:
Actually it works because it then uses the `TruffleObject` case in `com.oracle.truffle.regex.literal.LiteralRegexExecNode#doTruffleObject`.
That's not optimal for the same reason the `exec()` above is not optimal, because it can't use any intrinsic on the `byte[]`, it just treats it as an interop array of bytes.

This will be solved when we adopt TruffleString (in progress, but lots of work).
So might be fine to merge before, although I guess it might slower for some Regexps than the current code on master, which can use intrinsics on the byte[].

Given the code in LiteralRegexExecNode I wonder if it wouldn't be better to detect `Env#asGuestValue(bytes);` explicitly, rather than having execBytes, etc. Then it would correctly recognize it's a wrapped byte[] here.

Reply:
I'd like to get rid of `byte[]` inputs instead, since wrapping them in `TruffleString` objects allows for much better optimizations with `codeRange`.

Reply:
Yeah of course once TruffleRuby will have adopted TruffleString there won't be a need for the `byte[]` inputs anymore AFAIK.

Reply:
As suggested, we'll modify our interaction with TRegex to use TruffleString as that gets adopted more widely in TruffleRuby. Is there anything else to do in the meanwhile to get this merged?

Reply:
I'd like to wait for a indexed frame slots fix first if that's OK (should be soon), as this PR might change peak performance for some benchmarks and I wouldn't want to mix both.

Also this PR might actually regress performance because instrinsics and e.g. vectorized operations cannot be used with this PR for `match?`, but were before (at the cost of allocating the MatchData, which is probably a much smaller gain than vectorized operations in some usages).

Reply:
That's fine. It looked like it was being held up on a decision yet to be made. As long as there's a reason for it to still be open, that's fine with me.

Reply:
I'll integrate it now, it's always possible to run benchmarks on older commits/on a branch and this PR is also rather unlikely to regress many benchmarks.